utils/api_util.py

```diff
@@ -0,0 +1,139 @@
+
+            log.info("Failed data saved successfully.")
+
+
+async def _insert_df(df, url, session: ClientSession, sem):
+    """Wrapper call for API call to insert data"""
+    objs = df.to_dict(orient='records')
+    tasks = []
+    for obj in objs:
+        task = asyncio.ensure_future(_add_data(obj=obj, session=session, url=url, sem=sem))
+        tasks.append(task)
+    await asyncio.gather(*tasks)
+
+    log.info("Record insertion completed.")
+
+
+async def _add_data(obj, session: ClientSession, url, sem):
+    """Call API server and insert data."""
+    async with sem, session.post(url, json=obj) as response:
+        log.info('Got response {} for {}'.format(response.status, obj["url"]))
+        if response.status != 200:
+            failed_to_insert.append(obj["url"])
+            log.error('Error response {}, msg {}, data: {}'
+                      .format(response.status, await response.text(), json.dumps(obj)))
+
+
+def _get_status_type(status: str) -> str:
+    """Convert status to uppercase so API endpoint can understand the same."""
+    if status.lower() in ['opened', 'closed', 'reopened']:
+        return status.upper()
+    else:
+        return "OTHER"
+
+
+def _get_probabled_cve(cve_model_flag: int) -> bool:
+    """Get Ptobable CVE flag based on cve_model_flag."""
+    return True if cve_model_flag is not None and cve_model_flag == 1 else False
+
+
+def _update_df(df: pd.DataFrame) -> pd.DataFrame:
+    """Update few property of the dataframe to make it work with API sevrer."""
+    df.loc[:, "ecosystem"] = df['ecosystem'].str.upper()
+    df.loc[:, "status"] = df.apply(lambda x: _get_status_type(x['status']), axis=1)
+    df.loc[:, "probable_cve"] = df.apply(lambda x: _get_probabled_cve(x['cve_model_flag']), axis=1)
+
+    return df.where(pd.notnull(df), None)
+
+
+def _get_triage_subdir(start_time, end_time):
+    return cc.NEW_TRIAGE_SUBDIR.format(stat_time=start_time.format("YYYYMMDD"),
+                                                end_time=end_time.format("YYYYMMDD"))
+
+
+
+async def _update_and_save(start_time, end_time, cve_model_type: str, s3_upload: bool, ecosystem: str):
+    """Save probable cve data to db via api server."""
+    triage_subdir = _get_triage_subdir(start_time, end_time)
+    df = _read_probable_cve_data(triage_subdir, cve_model_type, s3_upload, ecosystem)
+    if len(df) != 0:
+
+        log.info("PCVE data count :{count}".format(count=str(df.shape[0])))
+        updated_df = _update_df(df)
+        log.info("Update df completed")
+
+        sem = asyncio.BoundedSemaphore(cc.DATA_INSERT_CONCURRENCY)
+
+        async with ClientSession() as session:
+            await _insert_df(updated_df, API_FULL_PATH, session, sem)
+
+        return updated_df
+
+    else:
+        log.info("No PCVE records to insert for {}".format(ecosystem))
+        return df
+
+
+def save_data_to_db(start_time, end_time, cve_model_type: str, s3_upload: bool, ecosystem: str):
```

Comment:
I still see that this function violates the SRP rule. If I understand correctly, it does inserts local data or S3 data into graphdb. I prefer the following approach,

```
def save_data_to_db(df: pd.DataFrame, ecosystem: str):
```
save_data_to_db can take `df` argument which could be either based inmemory or S3 based. Rest of the logic should be handled outside of `save_data_to_db`

Reply:
Now reading data outside of save_data_to_db function as suggested and passing dataframe to it.